### ISSUE TITLE
Un-expected behaviour for 'cabal new-build' command.

### ISSUE THREAD

**Issue by erikd** [haskell/cabal#4524]
I using git commit 967a64e3 (TueMay 16 22:17:46 2017 +0200 ) 

If I run `cabal new-build` in a directory with a cabal file I get:
```
> cabal new-build
cabal: No targets given and there is no package in the current directory. Use
the target 'all' for all packages in the project or specify packages or
components by name or location. See 'cabal build --help' for more details on
target options.
```
but there very definitely *is* a cabal file in the current directory, and `cabal build` at least tries to build it but can't because there's no sandbox.

**Comment by 23Skidoo** [comment]
Do you have a test case that we could use to reproduce this?

**Comment by erikd** [comment]
I'll whip up a simple test case.

**Comment by erikd** [comment]
Never mind, talking to @hvr on IRC sorted it out.

I had a `cabal.project` file containing:
```
packages: lib/Foor lib/Bar
```
but did not explicitly mention `.` , the current directory.

I suspect what's needed is a warning if `cabal new-build` is run and  there is a cabal file and a `cabal.project` file in the current directory but the `cabal.project` file doesn't include `.` in the `packages` field.

**Comment by ezyang** [comment]
I'm OK with adding a warning, but if `.` is omitted on purpose, how is a user supposed to suppress the warning?

**Comment by erikd** [comment]
@ezyang In what circumstances would someone have a `cabal.project` file with `,` and a cabal file in the same directory and not want this warning when running `cabal new-build`.

**Comment by ezyang** [comment]
If there really is no such circumstance, then new-build should always include `.` when it has a cabal file. If there is a circumstance, then they would not want this warning ;)

**Comment by erikd** [comment]
Discussed this with @hvr on IRC. He and I came to the conclusion that there should be no implicit additions the `cabal.project` `packages` field, only explicit.

If `.` needs to be explicitly added, then this warning makes sense.

**Comment by ezyang** [comment]
Go for it.

**Comment by cj285** [comment]
We hit this error message when relocating a project folder to a new directory ... deleting the dist-newstyle directory seemed to resolve it, but not obvious from the error message

**Comment by hvr** [comment]
Don't we just need to modify the error message slightly in? C.f. (boldfaced the incorrect and misleading part)

> cabal: No targets given and **there is no package in the current directory**. Use the target 'all' for all packages in the project or specify packages or components by name or location. See 'cabal build --help' for more details on target options.

So that error should change to something different in case there is a cloest `.cabal` file covering the current folder which isn't referred to by the `cabal.project(.local)` configuration. E.g.

> cabal: No targets given and **but there is a package in the current directory which isn't referred to in your `cabal.project` configuration**. Use the target 'all' for all **registered** packages in the project or specify packages or components by name or location. See 'cabal build --help' for more details on target options.

**Comment by quasicomputational** [comment]
FWIW I've been bitten by this when writing `cabal-testsuite` tests, and I've seen at least one other instance as well. I think @hvr's message change is sensible, but we should also warn when there's a `.cabal` file in any of the parent directories up to the project root but which isn't listed in `cabal.project`, even if the build otherwise looks kosher.

**Comment by hvr** [comment]
@quasicomputational while it may sound good at first, that's computationally more effort (and I'm not sure how realistic the corner-case you want to warn about actually is -- can you provide a concrete example you're worried about?); as you'd have to always check *all* parent folders; this would require to modify the lookup logic to do more work just for the sake of warning. Whereas, what I suggested is something that is covered by the mandatory lookup logic anyway (i.e. searching for the closest `.cabal` file + searching for the closest `cabal.project` file).

**Comment by quasicomputational** [comment]
Sorry, maybe I phrased that badly. The specific case that bit me was that I'd added a new test to the testsuite, with its own `.cabal` file, and the test driver used `cabal new-build all` (IIRC). But then that ignored the `.cabal` file next to it, and proceeded to try building `cabal-install`, `Cabal`, etc, because it found Cabal's `cabal.project` quite a few directories up! Finding the nearest `.cabal` and the nearest `cabal.project`, and then complaining if that `.cabal` isn't in `cabal.project` would have saved me puzzlement.

**Comment by hvr** [comment]
@quasicomputational ok, but then the heuristic I mentioned in my comment, i.e.

> So that error should change to something different in case there is a cloest .cabal file covering the current folder which isn't referred to by the cabal.project(.local) configuration ...

seems to suffice for your scenario, no?

**Comment by quasicomputational** [comment]
Yes, the heuristic is fine and I meant to agree with you about it; sorry for the non-clarity. What I want is for a warning to apply even if there are targets and `new-build` is otherwise happy. The discussion was about an error, but the problem I ran into was that there *wasn't* an error and I very much would have appreciated a hint that the `.cabal` file wasn't being used at all.
